### DISCUSSION TITLE
Separate (underlying) datachannels for unreliable and reliable messages

### DISCUSSION THREAD

**Discussion by reececomo** [geckosio/geckos.io#263] [Ideas]
I think Geckos.io would benefit from creating two separate datachannels per connection, one for unreliable and one for reliable.

That would allow certain messages to leverage the existing (super fast) reliable mode at the network layer instead of reimplementing it at the application layer.

```ts
// UDP-like channel for realtime data (e.g. binary, serialized game state packets).
const unreliable = RTCPeerConnection.createDataChannel("unreliable", {
    ordered: false,
    maxRetransmits: 0
});

// TCP-like channel for reliable data (e.g. chat message, ad-hoc game state events).
const reliable = RTCPeerConnection.createDataChannel("reliable", {
    ordered: true, // optional?
    maxRetransmits: 10 // Use constant
});
```

Edit: Moved to https://github.com/geckosio/geckos.io/issues/269

**Comment by eltonnobrega** [comment]
Hello. How are you?

Checking the code here, the answer is no. The gecko.io server receives a request and just creates an unreliable, unordered connection (default of @param options.ordered is false).
You can send Reliable Messages using the reliable option of emit(reliable: true). It uses one form of reliability, but there are many.
A good read are the articles by Glenn Fiedler(https://gafferongames.com/#posts).

I thought about using websocket for reliable connections, but a simplistic ordered reliable udp version will be faster.

Your idea of separating channels is good too, for example, the construct3 multiplayer plugin (where I create my games) makes 3 channels: unreliable and unordered, reliable and unordered, and reliable ordered.

**Reply by reececomo** [reply]
> Checking the code here, the answer is no.

My question was actually intended to be rhetorical - just a polite way of saying "Hey, I think we should take advantage of WebRTC's existing transport-layer reliable mode and expose a second data channel for reliable packets".

The "reliable" option implemented in geckos.io is not particularly reliable, and it's solving at the application layer - whereas WebRTC (node-datachannel) already provides a better reliable mode at the transport layer.

**Reply by eltonnobrega** [reply]
Hey how's it going? She was reading more about the subject. First, I must say that the SCTP protocol is well built and even configured as reliable and orderly it appears to be faster than TCP. I found a good explanation of how to configure channels here: https://stackoverflow.com/questions/54292824/webrtc-channel-reliability.

1- Creates a reliable & ordered channel, like TCP but message based instead of stream:
```
RTCPeerConnection.createDataChannel("label", {
     ordered: true
});
```
2 - Creates a reliable but unordered channel (should maxRetransmits or maxPacketLifeTime be specified too to achieve reliability?)
```
RTCPeerConnection.createDataChannel("label", {
         ordered: false
});
```
3 - Creates an unreliable and unordered channel, like UDP
```
RTCPeerConnection.createDataChannel("label", {
     ordered: false,
     maxRetransmits: 0
});
```

I make games using construct3. I saw that their multiplayer plugin creates precisely 3 data channels for the developer to decide which one to use. They even give some tips on which one is best for what you want to do. https://www.construct.net/en/tutorials/multiplayer-tutorial-concepts-579/construct-multiplayer-plugin-10 (RELIABILITY MODES)

The most reliable mode is Reliable ordered. This sends a message that is tracked: if it gets dropped by the network, it will re-send it until it verifies that it arrives at the destination. It also orders messages, so they are guaranteed to arrive in the same order they were sent. This is useful, but is not the fastest: if one message is dropped and held up, it will hold up all subsequent messages until that message gets through. It also means individual messages may take multiples of the latency to arrive since it must wait at least the round-trip time to know if the message has arrived. This is generally only suitable for important low-frequency messages like chat messages.

The next fastest is Reliable unordered. This sends a message that is tracked, and will also re-send it until it verifies it arrives at the destination. However messages are allowed to arrive in any order. This prevents a single held up message from holding up all the subsequent messages - it can simply arrive late, after other messages that were originally sent after it. However individual messages can still take a multiple of the latency to arrive. This is suitable for gameplay events that must arrive but are unrelated to one another, like a door opening or explosion occurring.

The fastest mode is Unreliable. This sends a single message and forgets about it. If it's dropped, the message will simply never arrive at the destination. It can also arrive in a different order to the sequence it was originally sent in, and even possibly in duplicate. Generally this is suitable for regular messages, where it is more important that the message arrives as soon as possible than whether it arrives at all. If it is dropped, it is likely to be followed up quickly by the next message with new data, so it doesn't matter. Note the engine internally uses this mode for synced objects with built-in interpolation and compensation features, so don't try to replicate that functionality with this mode.

So, the conclusion I see is that it would be really good if the 3 channels were created on geckos.io:

```
     // DataChannel configuration reliable and ordered
     const dc_config_reliable_ordered: DataChannelInitConfig = {
       ordered: true
     }
     const label_reliable_ordered = label+"_reliable_ordered";
     const dc_reliable_ordered = await promiseWithTimeout(createDataChannel(pc, label_reliable_ordered , dc_config_reliable_ordered), 2000)
     connection.channel_reliable_ordered = new CreateDataChannel(connection, dc_reliable_ordered , this.options, userData)

     // DataChannel configuration reliable and unordered
     const dc_config_reliable_unordered: DataChannelInitConfig = {
       ordered: false
     }
     const label_reliable_unordered = label+"_reliable_unordered";
     const dc_reliable_unordered = await promiseWithTimeout(createDataChannel(pc, label_reliable_unordered, dc_config_reliable_unordered), 2000)
     connection.channel_reliable_unordered = new CreateDataChannel(connection, dc_reliable_unordered , this.options, userData)

     // DataChannel configuration unreliable and unordered
     const dc_config_unreliable_unordered: DataChannelInitConfig = {
       ordered: false,
       maxRetransmits: 0
     }
     const label_unreliable_unordered = label+"_unreliable_unordered";
     const dc_unreliable_unordered = await promiseWithTimeout(createDataChannel(pc, label_unreliable_unordered, dc_config_unreliable_unordered), 2000)
     connection.channel_unreliable_unordered = new CreateDataChannel(connection, dc_unreliable_unordered , this.options, userData)
```

**Reply by reececomo** [reply]
Yep, that is roughly what the original request asks for 😄

I don't need the "reliable but unordered" channel myself, but leveraging separate UDP-like and a TCP-like channels on the same WebRTC connection would be extremely useful.

**Reply by eltonnobrega** [reply]
How are you?
Reading a little about the subject, I was able to create a simple server just using node-datachannel and better understand how these libraries work.

In geckos itself, it is possible to open other data channels, as the peer connection is already made. Here is an example of how to open a channel as you want.

Server:
```
import geckos from '@geckos.io/server'

const io = geckos()

io.listen() // default port is 9208

io.onConnection(channel => {

	const peerConnection = channel.webrtcConnection.peerConnection;

	const dc_config_reliable_ordered: DataChannelInitConfig = {
		ordered: true
	}
	const dc_reliable_ordered = peerConnection.createDataChannel("reliable_ordered", dc_config_reliable_ordered);
	
	dc_reliable_ordered.onOpen(() => {
		dc_reliable_ordered.sendMessage('connected');
		dc_reliable_ordered.onMessage(msg => {
			console.log(`got ${msg} in default dc_reliable_ordered`)
			dc_reliable_ordered.sendMessage('msg pong: ' + msg);
		});
	});
	
	channel.on('chat message', data => {
		console.log(`got ${data} from "chat message" in default channel`)
	})
})
```

Client:
```
const channel = geckos({
	url: 'http://localhost'
})

channel.onConnect(error => {
	if (error) {
		console.error(error.message)
	return
	}
	
	channel_reliable_ordered = null;
	channel.peerConnection.localPeerConnection.ondatachannel = ((e) => { //RTCDataChannelEvent
		channel_reliable_ordered = e.channel;
		channel_reliable_ordered.onmessage = (e) => { //MessageEvent
			console.log(e);
		}
	});
	//channel_reliable_ordered.send("test");

	channel.on('chat message', data => {
		console.log(`You got the message ${data}`)
	})

	channel.emit('chat message', 'Hello')
})
```

**Reply by reececomo** [reply]
Yeah, adding it this way just misses the entire Geckosio abstraction layer. Doing this with JSON.stringify() and JSON.parse() with standard payloads is OK.

**Reply by yandeu** [reply]
I guess we could add this to the library.
I've opened a new issue.

**Reply by reececomo** [reply]
@eltonnobrega Just a note, the client needs to wait for the channel's `readyState` to be open otherwise it throws an `InvalidStateError` on send.

(For reference, I ran into this minor race condition on certain mobile browsers)

```ts
const reliableChannel = event.channel;

reliableChannel.onopen(() => {
  // Now we're ready.
  reliableChannel.send('hello')
});
```
